### PR TITLE
✨ Interleave intermediate assistant text with tool calls

### PR DESCRIPTION
**Pull request by thiesgerken**
Stacked on #231 (`feature/session-compaction`).

## Problem
Only a turn's **final** assistant text was ever persisted (one event at turn end). Tool calls/results are written as they happen, so any narration the model emitted *between* tool calls was swallowed. On reload, all tools collapsed into a single group with the answer dumped after them — order lost, intermediate text gone.

## Phase 1 — persist & interleave
- `agent/loop.py`: the stream handler buffers text parts and flushes a segment via a new `on_assistant_text` callback when a `ToolCallPart` starts. Trailing text (no tool after it) is the final answer and is never flushed.
- `turns.py`: `on_assistant_text` appends a `{role: assistant, content, partial: true}` event at its real position.
- `transcript.py`: `completed_event_turns` ignores `partial` events, keeping reset/fork/retry/compaction anchored to the final assistant answer.
- Frontend `done` handler finalizes **every** streaming segment (not just the last), so live matches reload.

## Phase 2 — metadata + actions on all messages
- Turn tooltip gains **"message i of n"** (per-turn bubble numbering).
- Copy + fork + reset exposed on all conversational messages.

## Phase 3 — event-level fork & reset
- Fork/reset cut at **any** message, not just turn boundaries. Every user prompt and assistant bubble (intermediate or final) is a branch point.
- `engine._sliced_transcript_for_event_cut`: a final-assistant cut reuses the fold-aware turn slice; a mid-turn (partial) or user-prompt cut keeps older turns compacted and **rebuilds only the cut turn verbatim from its events** — safe because the newest turn is never folded/tool-compacted. `reset_to_turn`/`fork_session` route through it; retry stays turn-level.
- Reset/fork at a **user prompt** keeps the conversation through that prompt and drops its response — amend and re-run with a follow-up.
- Frontend: messages carry their own event index; fork/reset use it directly (no turn snapping); reset slices optimistically at the clicked message.

## Tests
- `test_run_agent_turn_flushes_intermediate_text_before_tool_calls`, `test_completed_event_turns_ignores_partial_assistant_events`, `test_fork_at_partial_assistant_rebuilds_cut_turn`, `test_reset_to_user_message_cuts_after_prompt`.
- Backend 259 relevant tests green; ruff + pyrefly clean. Frontend 60 tests, tsc + eslint + i18n parity clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes session transcript persistence, fork/reset slicing, and live history projection—core chat paths—with substantial test coverage but non-trivial edge cases when turns overlap or history is mid-sync.
> 
> **Overview**
> **Intermediate assistant text** is no longer folded into the final answer only. The agent loop flushes narration when a tool call starts as **`partial` assistant events** in the event log; turn completion, compaction, and message counts treat only the **final** assistant message as the turn boundary.
> 
> The **chat UI** finalizes every streaming segment on `done`, optionally refetches history for timestamps/usage/**message i of n**, and exposes **copy/fork/reset** on user and assistant bubbles. **Fork and reset** cut at a message’s **event index** (including mid-turn partials and user prompts); the session engine slices events and **rebuilds the cut turn** from events when the cut isn’t a final assistant answer.
> 
> **Turn tooltips** add per-bubble position, **tok/s**, and **TTFT** from aggregated LLM request logs (decode time excludes tool gaps). i18n keys added for EN/DE.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 233e96dca155d51f6883cd1640ff260eecff3b39. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: frontend/src/components/chat-view.tsx [hunk 1]
```diff
@@ -1267,16 +1309,20 @@ export function ChatView({
...
+            if (lastStreamIdx === -1) {
               updated.push({ kind: "assistant", content: msg.content, finalStatus: msg.final_status });
+            } else {
+              for (let i = 0; i < updated.length; i++) {
+                if (updated[i].kind !== "streaming") continue;
+                updated[i] =
+                  i === lastStreamIdx
+                    ? { kind: "assistant", content: msg.content, finalStatus: msg.final_status }
+                    : { kind: "assistant", content: (updated[i] as { content: string }).content, partial: true };
+              }
```
[the remark this answers is not in the document]
Comment (thiesgerken) [reply frontend/src/components/chat-view.tsx]: Fixed in 6f8c0476. The final answer now replaces a streamed bubble only when it's the tail (no `tool_call` after it); otherwise the streamed bubbles stay `partial` narration and the answer is appended as its own message — matching the persisted events on reload.

File: src/carapace/agent/loop.py [hunk 2]
```diff
@@ -158,6 +178,7 @@ async def _stream_handler(_ctx: Any, events: Any) -> None:
         if before_llm_call is not None:
             before_llm_call()
         current_thinking_parts.clear()
+        current_text_parts.clear()
```
[the remark this answers is not in the document]
Comment (thiesgerken) [reply src/carapace/agent/loop.py]: Fixed in 6f8c0476. The approval loop now flushes buffered text (`_flush_text_segment`) before clearing for the next sub-run, so narration emitted after a deferred sub-run's last tool call is persisted instead of dropped.

File: frontend/src/components/chat-view.tsx [hunk 3]
```diff
@@ -1267,15 +1312,24 @@ export function ChatView({
...
+              lastStreamIdx !== -1
+              && !updated.slice(lastStreamIdx + 1).some((m) => m.kind === "tool_call");
+            for (let i = 0; i < updated.length; i++) {
+              if (updated[i].kind !== "streaming") continue;
+              updated[i] =
+                finalWasStreamed && i === lastStreamIdx
+                  ? { kind: "assistant", content: msg.content, finalStatus: msg.final_status }
+                  : { kind: "assistant", content: (updated[i] as { content: string }).content, partial: true };
+            }
+            if (!finalWasStreamed && (msg.content || lastStreamIdx === -1)) {
```
[the remark this answers is not in the document]
Comment (thiesgerken) [reply frontend/src/components/chat-view.tsx]: Fixed in 4d61ff3c. Dropped the truthiness guard — the final assistant bubble is now appended unconditionally when it wasn't streamed, matching the backend which always persists a final assistant event.

File: frontend/src/components/chat-view.tsx [hunk 4]
```diff
@@ -1267,25 +1313,51 @@ export function ChatView({
...
+          fetchHistory(server, token, sessionId)
+            .then((history) =>
+              setMessages((prev) => [
+                ...projectHistoryToMessages(history),
+                ...prev.slice(doneBaselineLen),
+              ]),
+            )
+            .catch(() => {
+              // Refetch failed — keep the live transcript; metadata fills in on the next reload.
+            });
```
[the remark this answers is not in the document]
Comment (thiesgerken) [reply frontend/src/components/chat-view.tsx:1359]: Fixed in 233e96dc. Added a monotonic sequence guard (`doneRefetchSeqRef`): each done bumps the id and the async response only applies if it's still the newest, so a slower earlier refetch can't clobber a newer turn's projection — latest-wins regardless of resolution order.
